File: data-structure/advanced/Two Array Problem.py
import sys, random, math
sys.setrecursionlimit(1_000_000)
input = sys.stdin.readline

# -------------------------
# Implicit Treap (rope)
# -------------------------
class Node:
    __slots__ = ("v", "p", "sz", "l", "r", "rev")
    def __init__(self, v):
        self.v = v
        self.p = random.getrandbits(31)
        self.sz = 1
        self.l = None
        self.r = None
        self.rev = False

def _sz(t): return t.sz if t else 0

def _pull(t):
    t.sz = 1 + _sz(t.l) + _sz(t.r)

def _push(t):
    if t and t.rev:
        t.rev = False
        t.l, t.r = t.r, t.l
        if t.l: t.l.rev ^= True
        if t.r: t.r.rev ^= True

def merge(a, b):
    if not a: return b
    if not b: return a
    _push(a); _push(b)
    if a.p > b.p:
        a.r = merge(a.r, b)
        _pull(a)
        return a
    else:
        b.l = merge(a, b.l)
        _pull(b)
        return b

def split(t, k):
    """split by first k elements: returns (a,b)"""
    if not t: return (None, None)
    _push(t)
    if _sz(t.l) >= k:
        a, b = split(t.l, k)
        t.l = b
        _pull(t)
        return (a, t)
    else:
        a, b = split(t.r, k - _sz(t.l) - 1)
        t.r = a
        _pull(t)
        return (t, b)

def build_treap(arr):
    # O(n) Cartesian build by random priorities
    st = []
    nodes = [Node(v) for v in arr]
    for nd in nodes:
        last = None
        while st and st[-1].p < nd.p:
            last = st.pop()
        if st:
            st[-1].r = nd
        nd.l = last
        st.append(nd)
    root = st[0] if st else None
    # fix sizes
    if root:
        stack = [(root, 0)]
        while stack:
            u, state = stack.pop()
            if not u: continue
            if state == 0:
                stack.append((u, 1))
                stack.append((u.r, 0))
                stack.append((u.l, 0))
            else:
                _pull(u)
    return root

def inorder_collect(t, out):
    stack = []
    cur = t
    while cur or stack:
        while cur:
            _push(cur)
            stack.append(cur)
            cur = cur.l
        cur = stack.pop()
        out.append(cur.v)
        cur = cur.r

# -------------------------
# Minimum Enclosing Circle
# -------------------------
EPS = 1e-10

def is_in_circle(c, p):
    cx, cy, r = c
    dx = p[0] - cx
    dy = p[1] - cy
    return dx*dx + dy*dy <= r*r + 1e-9

def circle_from_1(p):
    return (p[0], p[1], 0.0)

def circle_from_2(a, b):
    cx = (a[0] + b[0]) / 2.0
    cy = (a[1] + b[1]) / 2.0
    r = math.hypot(a[0] - b[0], a[1] - b[1]) / 2.0
    return (cx, cy, r)

def circle_from_3(a, b, c):
    ax, ay = a
    bx, by = b
    cx, cy = c
    d = 2.0 * (ax*(by-cy) + bx*(cy-ay) + cx*(ay-by))
    if abs(d) < EPS:
        return None
    ax2ay2 = ax*ax + ay*ay
    bx2by2 = bx*bx + by*by
    cx2cy2 = cx*cx + cy*cy
    ux = (ax2ay2*(by-cy) + bx2by2*(cy-ay) + cx2cy2*(ay-by)) / d
    uy = (ax2ay2*(cx-bx) + bx2by2*(ax-cx) + cx2cy2*(bx-ax)) / d
    r = math.hypot(ux-ax, uy-ay)
    return (ux, uy, r)

def min_enclosing_circle(points):
    pts = points[:]
    random.shuffle(pts)

    c = None
    for i, p in enumerate(pts):
        if c is not None and is_in_circle(c, p):
            continue
        c = circle_from_1(p)
        for j in range(i):
            q = pts[j]
            if is_in_circle(c, q):
                continue
            c = circle_from_2(p, q)
            for k in range(j):
                r = pts[k]
                if is_in_circle(c, r):
                    continue
                cc = circle_from_3(p, q, r)
                if cc is None:
                    # collinear: smallest circle from farthest pair
                    cands = [circle_from_2(p, q), circle_from_2(p, r), circle_from_2(q, r)]
                    best = None
                    for cand in cands:
                        if is_in_circle(cand, p) and is_in_circle(cand, q) and is_in_circle(cand, r):
                            if best is None or cand[2] < best[2]:
                                best = cand
                    c = best
                else:
                    c = cc
    return c

# -------------------------
# Operations
# -------------------------
def reverse_range(root, l, r):
    a, bc = split(root, l-1)
    b, c = split(bc, r-l+1)
    if b: b.rev ^= True
    return merge(a, merge(b, c))

def swap_segments_same_array(root, l1, r1, l2, r2):
    # Swap two disjoint segments in the SAME array (can have a gap)
    if l1 > l2:
        l1, l2 = l2, l1
        r1, r2 = r2, r1
    # assume disjoint: r1 < l2
    pre, rest = split(root, l1 - 1)
    seg1, rest = split(rest, r1 - l1 + 1)
    mid, rest = split(rest, l2 - r1 - 1)   # gap between seg1 and seg2 (could be empty)
    seg2, suf = split(rest, r2 - l2 + 1)
    return merge(pre, merge(seg2, merge(mid, merge(seg1, suf))))

def swap_between(a_root, b_root, l, r):
    a1, a_mid_suf = split(a_root, l-1)
    a_mid, a3 = split(a_mid_suf, r-l+1)
    b1, b_mid_suf = split(b_root, l-1)
    b_mid, b3 = split(b_mid_suf, r-l+1)
    a_root = merge(a1, merge(b_mid, a3))
    b_root = merge(b1, merge(a_mid, b3))
    return a_root, b_root

def extract_segment(root, l, r):
    pre, mid_suf = split(root, l-1)
    mid, suf = split(mid_suf, r-l+1)
    arr = []
    inorder_collect(mid, arr)
    root = merge(pre, merge(mid, suf))
    return root, arr

# -------------------------
# Main
# -------------------------
def main():
    n, q = map(int, input().split())
    A = list(map(int, input().split()))
    B = list(map(int, input().split()))

    a_root = build_treap(A)
    b_root = build_treap(B)

    out = []
    for _ in range(q):
        parts = list(map(int, input().split()))
        t = parts[0]

        if t == 1:
            # 1 p l r : reverse in array p (0=A, 1=B)
            p, l, r = parts[1], parts[2], parts[3]
            if p == 0:
                a_root = reverse_range(a_root, l, r)
            else:
                b_root = reverse_range(b_root, l, r)

        elif t == 2:
            # 2 p l1 r1 l2 r2 : swap two disjoint fragments (NOT necessarily consecutive)
            p, l1, r1, l2, r2 = parts[1], parts[2], parts[3], parts[4], parts[5]
            if p == 0:
                a_root = swap_segments_same_array(a_root, l1, r1, l2, r2)
            else:
                b_root = swap_segments_same_array(b_root, l1, r1, l2, r2)

        elif t == 3:
            # 3 l r : swap A[l..r] with B[l..r]
            l, r = parts[1], parts[2]
            a_root, b_root = swap_between(a_root, b_root, l, r)

        else:
            # 4 l r : MEC radius of points (A[i], B[i])
            l, r = parts[1], parts[2]
            a_root, xs = extract_segment(a_root, l, r)
            b_root, ys = extract_segment(b_root, l, r)
            pts = list(zip(xs, ys))
            if len(pts) <= 1:
                out.append("0.00")
            else:
                c = min_enclosing_circle(pts)
                out.append(f"{c[2]:.2f}")

    sys.stdout.write("\n".join(out))

if __name__ == "__main__":
    main()


# Timedout 

File: data-structure/advanced/Two Array Problem.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <math.h>
#include <string.h>

#define EPS 1e-12

// -------------------------
// Fast RNG (xorshift32)
// -------------------------
static uint32_t rng_state = 123456789u;
static inline uint32_t rng_u32() {
    uint32_t x = rng_state;
    x ^= x << 13;
    x ^= x >> 17;
    x ^= x << 5;
    rng_state = x;
    return x;
}
static inline int rng_int(int n) { // 0..n-1
    return (int)(rng_u32() % (uint32_t)n);
}

// -------------------------
// Implicit Treap
// -------------------------
typedef struct Node {
    int v;
    uint32_t pri;
    int sz;
    unsigned char rev;
    struct Node *l, *r;
} Node;

static inline int sz(Node *t) { return t ? t->sz : 0; }

static inline void pull(Node *t) {
    t->sz = 1 + sz(t->l) + sz(t->r);
}

static inline void push(Node *t) {
    if (t && t->rev) {
        t->rev = 0;
        Node *tmp = t->l;
        t->l = t->r;
        t->r = tmp;
        if (t->l) t->l->rev ^= 1;
        if (t->r) t->r->rev ^= 1;
    }
}

static Node* new_node(int v) {
    Node *n = (Node*)malloc(sizeof(Node));
    n->v = v;
    n->pri = rng_u32();
    n->sz = 1;
    n->rev = 0;
    n->l = n->r = NULL;
    return n;
}

static Node* merge(Node *a, Node *b) {
    if (!a) return b;
    if (!b) return a;
    push(a); push(b);
    if (a->pri > b->pri) {
        a->r = merge(a->r, b);
        pull(a);
        return a;
    } else {
        b->l = merge(a, b->l);
        pull(b);
        return b;
    }
}

// split by first k elements: left has k, right has rest
static void split(Node *t, int k, Node **a, Node **b) {
    if (!t) { *a = NULL; *b = NULL; return; }
    push(t);
    if (sz(t->l) >= k) {
        split(t->l, k, a, &t->l);
        pull(t);
        *b = t;
    } else {
        split(t->r, k - sz(t->l) - 1, &t->r, b);
        pull(t);
        *a = t;
    }
}

// Build treap in O(n) using a stack (Cartesian by priority)
static Node* build_treap(int *arr, int n) {
    Node **st = (Node**)malloc(sizeof(Node*) * n);
    int top = 0;
    for (int i = 0; i < n; i++) {
        Node *nd = new_node(arr[i]);
        Node *last = NULL;
        while (top > 0 && st[top-1]->pri < nd->pri) {
            last = st[--top];
        }
        if (top > 0) st[top-1]->r = nd;
        nd->l = last;
        st[top++] = nd;
    }
    Node *root = (top > 0) ? st[0] : NULL;

    // fix sizes with iterative postorder
    if (root) {
        Node **stack = (Node**)malloc(sizeof(Node*) * n * 2);
        unsigned char *state = (unsigned char*)malloc(sizeof(unsigned char) * n * 2);
        int sp = 0;
        stack[sp] = root; state[sp] = 0; sp++;
        while (sp) {
            Node *u = stack[sp-1];
            unsigned char s = state[sp-1];
            sp--;
            if (!u) continue;
            if (s == 0) {
                stack[sp] = u; state[sp] = 1; sp++;
                stack[sp] = u->r; state[sp] = 0; sp++;
                stack[sp] = u->l; state[sp] = 0; sp++;
            } else {
                pull(u);
            }
        }
        free(stack); free(state);
    }

    free(st);
    return root;
}

// inorder collect segment values into out[] (iterative), respecting lazy rev
static void inorder_collect(Node *t, int *out, int *idx) {
    Node **stack = (Node**)malloc(sizeof(Node*) * (t ? t->sz : 1));
    int sp = 0;
    Node *cur = t;
    while (cur || sp) {
        while (cur) {
            push(cur);
            stack[sp++] = cur;
            cur = cur->l;
        }
        cur = stack[--sp];
        out[(*idx)++] = cur->v;
        cur = cur->r;
    }
    free(stack);
}

// -------------------------
// Query ops
// -------------------------
static Node* reverse_range(Node *root, int l, int r) {
    Node *a, *bc, *b, *c;
    split(root, l-1, &a, &bc);
    split(bc, r-l+1, &b, &c);
    if (b) b->rev ^= 1;
    return merge(a, merge(b, c));
}

// swap two disjoint segments in same array (can have a gap). assumes non-overlap.
static Node* swap_segments(Node *root, int l1, int r1, int l2, int r2) {
    if (l1 > l2) { int tl=l1; l1=l2; l2=tl; int tr=r1; r1=r2; r2=tr; }
    // assume r1 < l2
    Node *pre, *rest, *seg1, *mid, *seg2, *suf;

    split(root, l1-1, &pre, &rest);
    split(rest, r1-l1+1, &seg1, &rest);
    split(rest, l2-r1-1, &mid, &rest);         // gap (maybe empty)
    split(rest, r2-l2+1, &seg2, &suf);

    return merge(pre, merge(seg2, merge(mid, merge(seg1, suf))));
}

static void swap_between(Node **A, Node **B, int l, int r) {
    Node *a1, *a2, *a3;
    Node *b1, *b2, *b3;
    split(*A, l-1, &a1, &a2);
    split(a2, r-l+1, &a2, &a3);

    split(*B, l-1, &b1, &b2);
    split(b2, r-l+1, &b2, &b3);

    *A = merge(a1, merge(b2, a3));
    *B = merge(b1, merge(a2, b3));
}

// extract values of segment [l..r] without changing treap
static void extract_segment(Node **root, int l, int r, int *out) {
    Node *pre, *mid_suf, *mid, *suf;
    split(*root, l-1, &pre, &mid_suf);
    split(mid_suf, r-l+1, &mid, &suf);
    int idx = 0;
    inorder_collect(mid, out, &idx);
    *root = merge(pre, merge(mid, suf));
}

// -------------------------
// Minimum Enclosing Circle
// -------------------------
typedef struct { double x, y; } Pt;
typedef struct { double x, y, r; } Circ;

static inline double dist2(Pt a, Pt b) {
    double dx = a.x - b.x, dy = a.y - b.y;
    return dx*dx + dy*dy;
}

static inline int in_circle(Circ c, Pt p) {
    double dx = p.x - c.x, dy = p.y - c.y;
    return dx*dx + dy*dy <= c.r*c.r + 1e-9;
}

static Circ circle1(Pt a) {
    Circ c; c.x = a.x; c.y = a.y; c.r = 0.0; return c;
}

static Circ circle2(Pt a, Pt b) {
    Circ c;
    c.x = (a.x + b.x) * 0.5;
    c.y = (a.y + b.y) * 0.5;
    c.r = sqrt(dist2(a, b)) * 0.5;
    return c;
}

static int circle3(Pt a, Pt b, Pt cpt, Circ *out) {
    double ax=a.x, ay=a.y, bx=b.x, by=b.y, cx=cpt.x, cy=cpt.y;
    double d = 2.0 * (ax*(by-cy) + bx*(cy-ay) + cx*(ay-by));
    if (fabs(d) < EPS) return 0;
    double ax2ay2 = ax*ax + ay*ay;
    double bx2by2 = bx*bx + by*by;
    double cx2cy2 = cx*cx + cy*cy;
    double ux = (ax2ay2*(by-cy) + bx2by2*(cy-ay) + cx2cy2*(ay-by)) / d;
    double uy = (ax2ay2*(cx-bx) + bx2by2*(ax-cx) + cx2cy2*(bx-ax)) / d;
    out->x = ux;
    out->y = uy;
    out->r = sqrt((ux-ax)*(ux-ax) + (uy-ay)*(uy-ay));
    return 1;
}

static void shuffle_pts(Pt *p, int n) {
    for (int i = n-1; i > 0; i--) {
        int j = rng_int(i+1);
        Pt tmp = p[i]; p[i] = p[j]; p[j] = tmp;
    }
}

static Circ min_enclosing_circle(Pt *pts, int n) {
    shuffle_pts(pts, n);
    Circ c = circle1(pts[0]);

    for (int i = 0; i < n; i++) {
        if (in_circle(c, pts[i])) continue;
        c = circle1(pts[i]);
        for (int j = 0; j < i; j++) {
            if (in_circle(c, pts[j])) continue;
            c = circle2(pts[i], pts[j]);
            for (int k = 0; k < j; k++) {
                if (in_circle(c, pts[k])) continue;
                Circ cc;
                if (circle3(pts[i], pts[j], pts[k], &cc)) {
                    c = cc;
                } else {
                    // collinear: pick smallest circle from pairs that covers all 3
                    Circ c1 = circle2(pts[i], pts[j]);
                    Circ c2 = circle2(pts[i], pts[k]);
                    Circ c3c = circle2(pts[j], pts[k]);
                    Circ best = c1;
                    if ((!in_circle(best, pts[i]) || !in_circle(best, pts[j]) || !in_circle(best, pts[k])) ||
                        (c2.r < best.r && in_circle(c2, pts[i]) && in_circle(c2, pts[j]) && in_circle(c2, pts[k])))
                        best = c2;
                    if (c3c.r < best.r && in_circle(c3c, pts[i]) && in_circle(c3c, pts[j]) && in_circle(c3c, pts[k]))
                        best = c3c;
                    c = best;
                }
            }
        }
    }
    return c;
}

// -------------------------
// Main
// -------------------------
int main() {
    int n, q;
    if (scanf("%d %d", &n, &q) != 2) return 0;

    int *A = (int*)malloc(sizeof(int) * n);
    int *B = (int*)malloc(sizeof(int) * n);
    for (int i = 0; i < n; i++) scanf("%d", &A[i]);
    for (int i = 0; i < n; i++) scanf("%d", &B[i]);

    Node *a_root = build_treap(A, n);
    Node *b_root = build_treap(B, n);

    free(A); free(B);

    // buffers for extraction
    int *xs = (int*)malloc(sizeof(int) * n);
    int *ys = (int*)malloc(sizeof(int) * n);
    Pt  *pts = (Pt*)malloc(sizeof(Pt) * n);

    for (int qi = 0; qi < q; qi++) {
        int t;
        scanf("%d", &t);
        if (t == 1) {
            int p, l, r;
            scanf("%d %d %d", &p, &l, &r);
            if (p == 0) a_root = reverse_range(a_root, l, r);
            else        b_root = reverse_range(b_root, l, r);
        } else if (t == 2) {
            int p, l1, r1, l2, r2;
            scanf("%d %d %d %d %d", &p, &l1, &r1, &l2, &r2);
            if (p == 0) a_root = swap_segments(a_root, l1, r1, l2, r2);
            else        b_root = swap_segments(b_root, l1, r1, l2, r2);
        } else if (t == 3) {
            int l, r;
            scanf("%d %d", &l, &r);
            swap_between(&a_root, &b_root, l, r);
        } else { // t == 4
            int l, r;
            scanf("%d %d", &l, &r);
            int len = r - l + 1;

            extract_segment(&a_root, l, r, xs);
            extract_segment(&b_root, l, r, ys);

            if (len <= 1) {
                printf("0.00\n");
                continue;
            }

            for (int i = 0; i < len; i++) {
                pts[i].x = (double)xs[i];
                pts[i].y = (double)ys[i];
            }

            Circ c = min_enclosing_circle(pts, len);
            printf("%.2f\n", c.r + 1e-9);
        }
    }

    free(xs);
    free(ys);
    free(pts);
    return 0;
}
